### PR TITLE
sched: provide an option for plugin developers to move pods to activeQ

### PR DESCRIPTION
**Pull request by Huang-Wei**
#### What type of PR is this?

/kind feature
/sig scheduling

#### What this PR does / why we need it:

In the ML/Batch fields, a series of Pods are usually considered as a unit that requires to be co-scheduled, or scheduled in a particular order (e.g., a spark job requires the driver to be running prior to executors). The current scheduling mechanics can make them work, but sometimes not efficient enough as different pods of a unit may be present in different internal queues and backoff-ed in different paces, and hence needs extra scheduling cycles to be scheduled. (see more explanation in the design doc)

This PR introduces a reserved CycleState (both key and data struct) and plumbs it into the scheduling cycle, so plugin developers can choose to move particular pods based on their requirements, proactively. The queued-up Pods will be moved to activeQ at the end of a successful scheduling cycle and binding cycle, in a best-efforts manner.

#### Which issue(s) this PR fixes:

Part of #100347.

#### Special notes for your reviewer:

In addition to the integration test, I also apply this patch to scheduler-plugins, and verified the integration performance of coscheduling plugin. The time reduced sharply from 400s~600s to ~17s:

Original test:

```
root@wei-dev:~/go/src/sigs.k8s.io/scheduler-plugins# for i in {1..20}; do go test -timeout 20m ./test/integration/... -run ^TestCoschedulingPlugin -count=1; done
ok      sigs.k8s.io/scheduler-plugins/test/integration  456.129s
ok      sigs.k8s.io/scheduler-plugins/test/integration  456.756s
ok      sigs.k8s.io/scheduler-plugins/test/integration  456.668s
ok      sigs.k8s.io/scheduler-plugins/test/integration  547.129s
ok      sigs.k8s.io/scheduler-plugins/test/integration  456.216s
ok      sigs.k8s.io/scheduler-plugins/test/integration  456.751s
ok      sigs.k8s.io/scheduler-plugins/test/integration  456.487s
ok      sigs.k8s.io/scheduler-plugins/test/integration  457.138s
ok      sigs.k8s.io/scheduler-plugins/test/integration  456.431s
ok      sigs.k8s.io/scheduler-plugins/test/integration  457.470s
...
```

With this PR and proactive moving pods logic:

```
root@wei-dev:~/go/src/sigs.k8s.io/scheduler-plugins# for i in {1..20}; do go test -timeout 20m ./test/integration/... -run ^TestCoschedulingPlugin -count=1; done
ok      sigs.k8s.io/scheduler-plugins/test/integration  16.867s
ok      sigs.k8s.io/scheduler-plugins/test/integration  16.367s
ok      sigs.k8s.io/scheduler-plugins/test/integration  16.836s
ok      sigs.k8s.io/scheduler-plugins/test/integration  16.842s
ok      sigs.k8s.io/scheduler-plugins/test/integration  16.363s
ok      sigs.k8s.io/scheduler-plugins/test/integration  16.441s
ok      sigs.k8s.io/scheduler-plugins/test/integration  16.506s
ok      sigs.k8s.io/scheduler-plugins/test/integration  16.970s
ok      sigs.k8s.io/scheduler-plugins/test/integration  17.014s
ok      sigs.k8s.io/scheduler-plugins/test/integration  16.739s
...
```

#### Does this PR introduce a user-facing change?

```release-note
Scheduler nows provides an option for plugin developers to move Pods to activeQ proactively.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [Design doc]: https://docs.google.com/document/d/1mdOJN9ceICMTh4sbSZezK5WuzH9OM8ou6IVuyN3CY0U/view
```

### CODE REVIEW

**Comment by Huang-Wei** [comment]
/assign @ahg-g @alculquicondor 
cc/ @yuanchen8911 @denkensk @cwdsuzhou

File: pkg/scheduler/scheduler.go [hunk 1]
```diff
@@ -667,6 +681,17 @@ func (sched *Scheduler) scheduleOne(ctx context.Context) {
 
 			// Run "postbind" plugins.
 			fwk.RunPostBindPlugins(bindingCycleCtx, state, assumedPod, scheduleResult.SuggestedHost)
+
+			// At the end of a successful binding cycle, move up Pods if needed.
+			if len(podsToMove.PodsMap) != 0 {
+				var movingPods []*v1.Pod
+				for _, pod := range podsToMove.PodsMap {
+					movingPods = append(movingPods, pod)
+					// Unlike the logic in scheduling cycle, we don't both deleting the entries
```
Comment (Huang-Wei) [inline pkg/scheduler/scheduler.go:690]: oops, both->bother. Will fix the typo along with other comments.

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 2]
```diff
@@ -77,6 +77,8 @@ type PreEnqueueCheck func(pod *v1.Pod) bool
 type SchedulingQueue interface {
 	framework.PodNominator
 	Add(pod *v1.Pod) error
+	// MoveUpPods moves given pods to activeQ in a best-efforts manner.
+	MoveUpPods(pods []*v1.Pod)
```
Comment (denkensk) [inline pkg/scheduler/internal/queue/scheduling_queue.go:81]: `MoveUpPods` may not be very easy to understand what he is going to do. Do you need to rename it like
`MoveAllToActiveQ`、`MoveAllToActiveQueue` or other better name？

File: pkg/scheduler/framework/interface.go [hunk 3]
```diff
@@ -105,6 +106,28 @@ const (
 	MaxTotalScore int64 = math.MaxInt64
 )
 
+// PodsToMoveKey is a reserved state key for stashing pods that
+// will be moved to activeQ at the end of a scheduling and binding cycle,
+// in a best-efforts manner.
+var PodsToMoveKey StateKey = "kubernetes.io/pods-to-move"
+
+// PodsToMove stores pods to be moved.
+type PodsToMove struct {
```
Comment (denkensk) [inline pkg/scheduler/framework/interface.go:115]: Can we consider using sync.map？

File: pkg/scheduler/scheduler.go [hunk 4]
```diff
@@ -608,6 +612,16 @@ func (sched *Scheduler) scheduleOne(ctx context.Context) {
 		return
 	}
 
+	// At the end of a successful scheduling cycle, pop and move up Pods if needed.
+	if len(podsToMove.PodsMap) != 0 {
```
Comment (denkensk) [inline pkg/scheduler/scheduler.go:616]: This place does not need to acquire lock under the normal logic.
But a little doubt：Is it possible that some other users will start a new go-routine in their plugins to R/W `podsToMove`?

**Review by denkensk — COMMENTED** [review]
Thanks so much. This optimization will significantly improve the efficiency of batch scheduling.
Some small comments.
@Huang-Wei

File: test/integration/scheduler/framework_test.go [hunk 5]
```diff
@@ -1941,6 +1944,141 @@ func TestPreemptWithPermitPlugin(t *testing.T) {
...
+				s.Unlock()
+			}
+			j.moveUpCalled = true
+		}
+	}
+}
+
+// This test simulates a typical spark job workflow.
+// - N executor pods are created, but kept pending due to missing the driver pod
+// - when the driver pod gets created and created, proactively move the executors to activeQ
```
Comment (chendave) [inline test/integration/scheduler/framework_test.go:2005]:
```suggestion
// - when the driver pod gets created, proactively move the executors to activeQ
```

File: pkg/scheduler/framework/interface.go [hunk 6]
```diff
@@ -105,6 +106,28 @@ const (
 	MaxTotalScore int64 = math.MaxInt64
 )
 
+// PodsToMoveKey is a reserved state key for stashing pods that
+// will be moved to activeQ at the end of a scheduling and binding cycle,
```
Comment (chendave) [inline pkg/scheduler/framework/interface.go:110]: this is really depending on when the pods are added.

File: pkg/scheduler/internal/queue/scheduling_queue.go (hunk 2 above)
Comment (chendave) [reply pkg/scheduler/internal/queue/scheduling_queue.go:81]: only the pods in the map will be moved, `MoveAll*` seems to me is a little confusion IMHO.

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 7]
```diff
@@ -77,6 +77,8 @@ type PreEnqueueCheck func(pod *v1.Pod) bool
 type SchedulingQueue interface {
 	framework.PodNominator
 	Add(pod *v1.Pod) error
+	// MoveUpPods moves given pods to activeQ in a best-efforts manner.
```
Comment (chendave) [inline pkg/scheduler/internal/queue/scheduling_queue.go:80]: explain what's the `given pods`, where it comes from and the rationale behind the scenes for better understanding of the motivation.

File: pkg/scheduler/scheduler.go [hunk 8]
```diff
@@ -608,6 +612,16 @@ func (sched *Scheduler) scheduleOne(ctx context.Context) {
...
+		var movingPods []*v1.Pod
+		for key, pod := range podsToMove.PodsMap {
+			movingPods = append(movingPods, pod)
+			delete(podsToMove.PodsMap, key)
+		}
+		sched.SchedulingQueue.MoveUpPods(movingPods)
```
Comment (ahg-g) [inline pkg/scheduler/scheduler.go:622]: change the signature to a receive a map to avoid the copy above.

File: pkg/scheduler/internal/queue/scheduling_queue.go (hunk 2 above)
Comment (ahg-g) [reply pkg/scheduler/internal/queue/scheduling_queue.go:81]: `MoveToActiveQueue` sounds good to me.

File: pkg/scheduler/framework/interface.go [hunk 9]
```diff
@@ -105,6 +106,28 @@ const (
...
+// in a best-efforts manner.
+var PodsToMoveKey StateKey = "kubernetes.io/pods-to-move"
+
+// PodsToMove stores pods to be moved.
```
Comment (ahg-g) [inline pkg/scheduler/framework/interface.go:114]: may be `PodsToActivate`?

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 10]
```diff
@@ -281,6 +283,34 @@ func (p *PriorityQueue) Run() {
...
+	return nil
+}
+
+func (p *PriorityQueue) MoveUpPods(pods []*v1.Pod) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	moved := false
+	for _, pod := range pods {
+		if err := p.add(pod); err == nil {
```
Comment (ahg-g) [inline pkg/scheduler/internal/queue/scheduling_queue.go:301]: What we want is to add the pod iif it was in backoff/unschedulable, using `add` means any pod could be added. What if the pod was deleted by the time we add it here?

File: pkg/scheduler/framework/interface.go [hunk 11]
```diff
@@ -105,6 +106,28 @@ const (
...
+type PodsToMove struct {
+	sync.Mutex
+	// PodsMap is keyed with namespaced pod name, and valued with the pod.
+	PodsMap map[string]*v1.Pod
```
Comment (ahg-g) [inline pkg/scheduler/framework/interface.go:118]: just `Map` or `Pods` or `Values`?

File: pkg/scheduler/internal/queue/scheduling_queue.go (hunk 2 above)
Comment (Huang-Wei) [reply pkg/scheduler/internal/queue/scheduling_queue.go:81]: My old plan was to add the function to FrameworkHandle, and hesitated to expose the internal details (activeQ) to the name. But now I'm implementing in a lightweight approach, and MoveXYZ only shows up in ScheduingQueue, so yes, `MoveToActiveQueue` is fine.

File: pkg/scheduler/framework/interface.go (hunk 3 above)
Comment (Huang-Wei) [reply pkg/scheduler/framework/interface.go:115]: Not quite a fan of that :)

File: pkg/scheduler/scheduler.go (hunk 4 above)
Comment (Huang-Wei) [reply pkg/scheduler/scheduler.go:616]: > Is it possible that some other users will start a new go-routine in their plugins to R/W `podsToMove`?

I cannot think of a reason why users would do that. Doing that means you choose to delay a piece of important information in a random manner (as you start a goroutine).

File: pkg/scheduler/framework/interface.go (hunk 9 above)
Comment (Huang-Wei) [reply pkg/scheduler/framework/interface.go:114]: sounds better.

File: pkg/scheduler/framework/interface.go (hunk 11 above)
Comment (Huang-Wei) [reply pkg/scheduler/framework/interface.go:118]: Map is good.

File: pkg/scheduler/internal/queue/scheduling_queue.go (hunk 7 above)
Comment (Huang-Wei) [reply pkg/scheduler/internal/queue/scheduling_queue.go:80]: updated.

File: pkg/scheduler/scheduler.go [hunk 12]
```diff
@@ -608,6 +612,16 @@ func (sched *Scheduler) scheduleOne(ctx context.Context) {
 		return
 	}
 
+	// At the end of a successful scheduling cycle, pop and move up Pods if needed.
+	if len(podsToMove.PodsMap) != 0 {
+		var movingPods []*v1.Pod
+		for key, pod := range podsToMove.PodsMap {
```
[the remark this answers is not in the document]
Comment (Huang-Wei) [reply pkg/scheduler/scheduler.go:618]: The sequence doesn't matter.

File: pkg/scheduler/scheduler.go [hunk 13]
```diff
@@ -667,6 +681,17 @@ func (sched *Scheduler) scheduleOne(ctx context.Context) {
 
 			// Run "postbind" plugins.
 			fwk.RunPostBindPlugins(bindingCycleCtx, state, assumedPod, scheduleResult.SuggestedHost)
+
+			// At the end of a successful binding cycle, move up Pods if needed.
```
[the remark this answers is not in the document]
Comment (Huang-Wei) [reply pkg/scheduler/scheduler.go:685]: `PostBind` is done in the binding cycle, i.e., in another goroutine.

File: pkg/scheduler/framework/interface.go (hunk 6 above)
Comment (Huang-Wei) [reply pkg/scheduler/framework/interface.go:110]: comments polished.

File: pkg/scheduler/internal/queue/scheduling_queue.go (hunk 10 above)
Comment (Huang-Wei) [reply pkg/scheduler/internal/queue/scheduling_queue.go:301]: Good point. We'd need to add a safety check here as users may pass in illegal pods.

File: pkg/scheduler/scheduler.go (hunk 8 above)
Comment (Huang-Wei) [reply pkg/scheduler/scheduler.go:622]: done.

**Comment by Huang-Wei** [comment]
Comments addressed. PTAL.

File: test/integration/scheduler/framework_test.go [hunk 14]
```diff
@@ -1941,6 +1944,141 @@ func TestPreemptWithPermitPlugin(t *testing.T) {
...
+		if err == nil {
+			if s, ok := c.(*framework.PodsToActivate); ok {
+				s.Lock()
+				for _, pod := range podsToMove {
+					namespacedName := fmt.Sprintf("%v/%v", pod.Namespace, pod.Name)
+					s.Map[namespacedName] = pod
+				}
+				s.Unlock()
+			}
+			j.moveUpCalled = true
```
Comment (chendave) [inline test/integration/scheduler/framework_test.go:1998]: Is it passible to capture the event of "ForceActivate" somewhere?

**Review by chendave — COMMENTED** [review]
looks good!

File: test/integration/scheduler/framework_test.go (hunk 14 above)
Comment (Huang-Wei) [reply test/integration/scheduler/framework_test.go]: In a running cluster, it can be captured by metrics {active: ForceActivate}:

```go
metrics.SchedulerQueueIncomingPods.WithLabelValues("active", ForceActivate).Inc()
```

But in a test, I think the local variable is good enough.

File: test/integration/scheduler/framework_test.go [hunk 15]
```diff
@@ -1941,6 +1944,141 @@ func TestPreemptWithPermitPlugin(t *testing.T) {
...
+func (j *JobPlugin) PreFilterExtensions() framework.PreFilterExtensions {
+	return nil
+}
+
+func (j *JobPlugin) PostBind(_ context.Context, state *framework.CycleState, p *v1.Pod, nodeName string) {
+	if _, ok := p.Labels["driver"]; !ok {
+		return
+	}
+
+	// If it's a driver pod, move other executor pods proactively to accelerating the scheduling.
```
[the remark this answers is not in the document]
Comment (Huang-Wei) [reply test/integration/scheduler/framework_test.go:1984]: done.

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 16]
```diff
@@ -301,6 +305,50 @@ func (p *PriorityQueue) Add(pod *v1.Pod) error {
...
 
+// Activate moves the given pods to activeQ iff they're in unschedulableQ or backoffQ.
+func (p *PriorityQueue) Activate(pods map[string]*v1.Pod) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	activated := false
+	for _, pod := range pods {
+		if p.activate(pod) {
+			activated = true
```
Comment (cwdsuzhou) [inline pkg/scheduler/internal/queue/scheduling_queue.go:316]: We should `break` here?

Or all/one of pods activate, we call `p.cond.Broadcast()`?

File: pkg/scheduler/scheduler.go [hunk 17]
```diff
@@ -608,6 +612,13 @@ func (sched *Scheduler) scheduleOne(ctx context.Context) {
 		return
 	}
 
+	// At the end of a successful scheduling cycle, pop and move up Pods if needed.
+	if len(podsToActivate.Map) != 0 {
```
Comment (ahg-g) [inline pkg/scheduler/scheduler.go:616]: shouldn't we run this even if we were not successful in scheduling the pod, not only on success? the failed to schedule pod may want another pod to scheduler first and so it may want to activate it.

Comment (Huang-Wei) [reply pkg/scheduler/scheduler.go:616]: It's debatable but my option is that "allowing the failed pod to activate pods" makes the mechanics complicated/chaotic and thus hard to use/debug, sometimes may generate a cycling dependency.

Take a group of pods as an example, some are "critical path" and some are not. I'd like the "critical" pod (e.g., a spark driver) to trigger the non-critical (executors) pods' activating; instead of the other way around. In terms of gang scheduling, although the implementation differs, an efficient way to do this is to let the Nth pod enters into waiting, while also activating other pods.

In other words, for a group of pods, I'd assume them to be inter-dependent in a unidirectional manner (instead of bidirectional).

File: pkg/scheduler/framework/interface.go [hunk 18]
```diff
@@ -105,6 +106,27 @@ const (
 	MaxTotalScore int64 = math.MaxInt64
 )
 
+// PodsToActivateKey is a reserved state key for stashing pods that will be
+// moved to activeQ if they're present in unschedulableQ or backoffQ.
```
Comment (ahg-g) [inline pkg/scheduler/framework/interface.go:110]: Can you update the comment to clarify that adding pods to this state doesn't guarantee that they get activated, and clarify when they will actually get activated.

File: pkg/scheduler/scheduler.go (hunk 17 above)
Comment (ahg-g) [reply pkg/scheduler/scheduler.go:616]: Ok, sounds a bit ad-hoc either way. In any case, lets clarify those semantics, I left a comment about that.

**Comment by alculquicondor** [comment]
@Huang-Wei how risky is it to merge this PR? Asking because of the code freeze tomorrow.

**Comment by Huang-Wei** [comment]
> @Huang-Wei how risky is it to merge this PR? Asking because of the code freeze tomorrow.

I'd say the risk is low as no default plugins are using the semantics, and hence no chance to trigger any potential issue.

In terms of its impact on out-of-tree plugins, as I mentioned in the description, it's a boost for co-scheduling efficiency, and I believe it's also valuable for other scenarios.

I will resolve https://github.com/kubernetes/kubernetes/pull/103383#discussion_r665049632, and then squash the commits.

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 19]
```diff
@@ -301,6 +305,50 @@ func (p *PriorityQueue) Add(pod *v1.Pod) error {
...
+		}
+	}
+
+	if activated {
+		p.cond.Broadcast()
+	}
+}
+
+func (p *PriorityQueue) activate(pod *v1.Pod) bool {
+	pInfo := p.newQueuedPodInfo(pod)
```
Comment (alculquicondor) [inline pkg/scheduler/internal/queue/scheduling_queue.go:326]: Aren't we losing backoff information?

File: pkg/scheduler/framework/interface.go [hunk 20]
```diff
@@ -105,6 +106,27 @@ const (
...
+var PodsToActivateKey StateKey = "kubernetes.io/pods-to-activate"
+
+// PodsToActivate stores pods to be activated.
+type PodsToActivate struct {
+	sync.Mutex
+	// Map is keyed with namespaced pod name, and valued with the pod.
```
Comment (alculquicondor) [inline pkg/scheduler/framework/interface.go:116]: why do you need the pod, if we already have it in the unschedulable/backoff queues?

File: pkg/scheduler/framework/interface.go (hunk 18 above)
Comment (alculquicondor) [reply pkg/scheduler/framework/interface.go:110]: I wonder why use the State instead of having a dedicated method in the framework?

File: pkg/scheduler/internal/queue/scheduling_queue.go (hunk 19 above)
Comment (Huang-Wei) [reply pkg/scheduler/internal/queue/scheduling_queue.go:326]: Thanks for the catch. Updated. PTAL.

File: pkg/scheduler/framework/interface.go (hunk 20 above)
Comment (Huang-Wei) [reply pkg/scheduler/framework/interface.go:119]: well, we only need a namespaced name here, but the current interface in unschedulable/backoff queues only accepts `pod`, so have to adapt to it; otherwise adjusting the interface isn't trivial.

File: pkg/scheduler/framework/interface.go (hunk 18 above)
Comment (Huang-Wei) [reply pkg/scheduler/framework/interface.go]: The rationale behind this is: the `toActivedPods` are transient and only effective within a scheduling/binding cycle. So manipulating the reserved CycleState make the logic straightforward, and consistent for both in-tree and out-of-tree plugins - they're using the same interface, no hidden API. While using a dedicated method (like `handle.ActivatePods()`), we have to still implement the API as, probably a cycle-specific data struct, and then make the semantics understood by plugin developers, which is not less intuitional IMO.

File: pkg/scheduler/framework/interface.go [hunk 21]
```diff
@@ -106,8 +106,11 @@ const (
 	MaxTotalScore int64 = math.MaxInt64
 )
 
-// PodsToActivateKey is a reserved state key for stashing pods that will be
-// moved to activeQ if they're present in unschedulableQ or backoffQ.
+// PodsToActivateKey is a reserved state key for stashing pods.
+// If the stashed pods are present in unschedulableQ or backoffQ，they will be
+// activated (i.e., moved to activeQ) in two phases:
+// - end of a successful scheduling cycle (will be cleared from `PodsToActivate` if activated)
+// - end of a successful binding cycle
```
Comment (alculquicondor) [inline pkg/scheduler/framework/interface.go:113]: if they succeed

**Review by alculquicondor — COMMENTED** [review]
ready for squash
